bbc/simorgh
# Send SSR errors from Next.js app

Overall changes
======
- Sends `SERVER_SIDE_REQUEST_FAILED` logs when the initial server render fails
- This should mimic the behaviour of the Express app when data fetching works, but the actual server render fails
- Moves the `SERVER_SIDE_RENDER_REQUEST_RECEIVED` log to the `_document.tsx` entry point as well, as this is a more accurate reflection of the initial server side request

Code changes
======
- Adds SSR failure logs to the initial server render in the Next.js app
- Moves `SERVER_SIDE_RENDER_REQUEST_RECEIVED` out of `getServerSideProps` and into the `getInitialProps` of the `_document.tsx` file, which is the first entry point into the application
- Adds function to remove sensitive headers from the logs
- Adds function to derive the page type to add to the logs

Testing
======
1. Pull down this branch locally
2. Run the Next.js app and force the render to fail. You can do this by adding in some deliberately broken logic, for example in the `LivePageLayout.tsx` file add the following:
```
  const myobj = { some: 'val' };

  // @ts-expect-error - test
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const myval = myobj.some[0].blocks.model;
```
This will cause the app to crash as its trying to access a property on an object that doesn't exist
3. Check the terminal logs and confirm that the `500` error is logged. It should contain something like: `"pageType":"live","status":500,"url":"/burmese/live/cx8jgp1zxylt?renderer_env=live"},"metadata":{"event":"ssr_request_failed","file":"pages/_document.js","label":"NEXTJS"},"timestamp":"2024-04-16 10:44:36.115"}`


Helpful Links
======
_Add Links to useful resources related to this PR if applicable._

[Coding Standards](https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/README.md)

[Repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)
